src/pydrex/core.py

```diff
@@ -0,0 +1,634 @@
+
+    [1]: https://doi.org/10.1016/S0012-821X(01)00356-9
+    [2]: https://doi.org/10.1111/j.1365-246X.2004.02308.x
+
+    """
+    n_elements = orientations.shape[0]
+    rotation_rates = np.empty((n_elements, 3, 3))
+    for element_index in range(n_elements):
+        # TODO: Make sure that orientations[element_index] is only a pointer, not a copy.
+        slip_rates, slip_indices = _get_slip_rates(
+            orientations[element_index],
+            strain_rate,
+            rrss,
+            stress_exponent,
+        )
+        deformation_rate, slip_rate_softest = _get_deformation_rate(
+            orientations[element_index],
+            slip_rates,
+            rrss,
+            velocity_gradient,
+        )
+
+        # Spin vector for the volume element, see eq. 3 in Fraters 2021.
+        spin_vector = np.empty(3)
+        for j in range(3):
+            r = (j + 1) % 3
+            s = (j + 2) % 3
+            spin_vector[j] = (
+                velocity_gradient[s, r]
+                - velocity_gradient[r, s]
+                - (deformation_rate[s, r] - deformation_rate[r, s]) * slip_rate_softest
+            ) / 2
+
+        # Calculate rotation rate, see eq. 9 Kaminski & Ribe (2001).
+        for p in range(3):
+            for q in range(3):
+                for r in range(3):
+                    for s in range(3):
+                        rotation_rates[element_index, p, q] += (
+                            _tensors.PERMUTATION_SYMBOL[q, r, s]
+                            * orientations[element_index, p, s]
+                            * spin_vector[r]
+                        )
+
+    return rotation_rates, slip_rates, slip_indices, slip_rate_softest
+
+
+@nb.njit(fastmath=True)
+def get_strain_energy_residuals(
+    vol_dist,
+    slip_rates,
+    slip_indices,
+    slip_rate_softest,
+    rrss,
+    stress_exponent,
+    nucleation_efficiency,
+):
+    """Calculate strain energy residual for each volume element.
+
+    Args:
+        `vol_dist` (NxD array) — present volume distribution (discretized)
+        `slip_rates` (array) — slip rates relative to slip rate on softest slip system
+        `slip_indices` (array) — indices that sort the RRSS by increasing slip rate activity
+        `slip_rate_softest` (float) — slip rate on the softest (most active) slip system
+        `stress_exponent` (float) — exponent for the stress dependence of dislocation density
+        `nucleation_efficiency` (float) — parameter controlling grain nucleation
+
+    """
+    n_elements = len(vol_dist)
+    strain_energies = np.zeros(n_elements)
+    for element_index in range(n_elements):
+        # Dimensionless dislocation density for each slip system.
+        # See eq. 16 Fraters 2021.
+        # NOTE: Mistake in eq. 11, Kaminski 2004: spurrious division by strain rate scale.
+        if rrss == _fabric.RRSS_OLIVINE_A:
+            for i in slip_indices[1:]:
```

Comment:
This is equivalent to line 371 in PyDRex.py. There is some discussion in the Fraters paper that the fourth index should be used as well. However, the fourth component will be zero as long as we are using the olivine RRSS with `np.inf` for the fourth value (Fraters also use this). The original DRex code uses `1e60` and then leaves out the fourth component in this sum, so I'm thinking they meant for it to be effectively 1/infinity.